examples/fine-tuning/osft/osft-example.ipynb

```diff
@@ -0,0 +1,1138 @@
+    "    \"save_final_checkpoint\": True,\n",
+    "    \"checkpoint_at_epoch\": False,\n",
+    "\n",
+    "    # Please note that the distributed training parameters are removed because they are\n",
+    "    # delegated to Kubeflow Trainer\n",
+    "}\n",
+    "\n",
+    "print(\"⚙️  Training Hyperparameters\")\n",
+    "print(\"=\" * 50)\n",
+    "print(json.dumps(params, indent=4))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Define training function\n",
+    "\n",
+    "We need to define our training function which relies on OSFT algorithm."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "def osft_train_fn(params):\n",
+    "    import os, time, json, random, sys, logging\n",
+    "    from contextlib import redirect_stdout, redirect_stderr\n",
+    "    from datasets import load_dataset\n",
+    "    from training_hub import osft\n",
+    "    from io import StringIO\n",
+    "    import glob\n",
+    "\n",
+    "    # Model and data\n",
+    "    BASE_MODEL            = params.get(\"base_model\", \"Qwen/Qwen2.5-1.5B-Instruct\")\n",
+    "    DATASET_PATH          = params.get('dataset_path', \"/mnt/shared/table-gpt-data/train/train_All_5000.jsonl\")\n",
+    "    CHECKPOINTS_PATH      = params.get(\"checkpoints_path\", \"/mnt/shared/checkpoints-logs-dir\")\n",
+    "\n",
+    "    # OSFT-specific\n",
+    "    UNFREEZE_RANK_RATIO   = float(params.get(\"unfreeze_rank_ratio\", 0.25))\n",
+    "\n",
+    "    # Training parameters\n",
+    "    NUM_EPOCHS            = int(params.get(\"num_epochs\", 1))\n",
+    "    BATCH_SIZE            = int(params.get(\"batch_size\", 128))\n",
+    "    LEARNING_RATE         = float(params.get(\"learning_rate\", 5e-6))\n",
+    "    MAX_SEQ_LEN           = int(params.get(\"max_seq_len\", 8192))\n",
+    "    MAX_TOKENS_PER_GPU    = int(params.get(\"max_tokens_per_gpu\", 64000))\n",
+    "\n",
+    "    # Data processing \n",
+    "    DATA_OUTPUT_PATH      = params.get(\"data_output_path\", \"/dev/shm\")\n",
+    "    WARMUP_STEPS          = int(params.get(\"warmup_steps\", 0))\n",
+    "\n",
+    "    # Optimization\n",
+    "    USE_LIGER             = bool(params.get(\"use_liger\", True))\n",
+    "    SEED                  = int(params.get(\"seed\", 42))\n",
+    "    LR_SCHEDULER          = params.get(\"lr_scheduler\", \"cosine\")\n",
+    "\n",
+    "    # Checkpointing\n",
+    "    SAVE_FINAL_CHECKPOINT = params.get(\"save_final_checkpoint\", True)\n",
+    "    CHECKPOINT_AT_EPOCH   = params.get(\"checkpoint_at_epoch\", False)\n",
+    "    \n",
+    "    os.makedirs(CHECKPOINTS_PATH, exist_ok=True)\n",
+    "    LOGFILE = os.path.join(CHECKPOINTS_PATH, \"train.log\")\n",
+    "\n",
+    "    # Let's define some helper functions for checkpoint management.\n",
+    "    def find_most_recent_checkpoint(output_dir):\n",
```

Comment:
I think it might be better for this to be defined outside the training function - post training, not on the training pod. In this scenario, this will run on every training node right?

Reply:
It does run on every node yes and it's only to provide this output log:
`print(f"📁 Final model checkpoint: {final_checkpoint}")`

So if we remove it from training func I can also remove that log, wdyt would be better here?

Reply:
Thanks Brian